File: webhook-payload-not-verified/node/verify-webhook-secret.js
/**
 * Classify BigCommerce webhook requests so a handler never trusts an unverified payload.
 *
 * BigCommerce webhook callbacks carry a hash field, but BigCommerce has never published
 * a supported formula for validating it, so its presence is not real verification. The
 * documented safeguard is the optional headers object you set when creating a hook with
 * POST /v3/hooks, which BigCommerce echoes back on every callback as a shared secret.
 * This scans hooks for a missing secret, provisions one when confirmed missing, and
 * exposes a pure classification function a receiver can use before ever acting on the
 * payload. Run the scan on a schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/bigcommerce/webhook-payload-not-verified/
 */
import { randomBytes, timingSafeEqual } from "node:crypto";
import { pathToFileURL } from "node:url";

const STORE_HASH = process.env.BIGCOMMERCE_STORE_HASH || "example";
const TOKEN = process.env.BIGCOMMERCE_ACCESS_TOKEN || "dummy_token";
const BASE = `https://api.bigcommerce.com/stores/${STORE_HASH}/`;
const HEADER_NAME = process.env.WEBHOOK_SECRET_HEADER_NAME || "X-Webhook-Secret";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

function constantTimeEqual(a, b) {
  const bufA = Buffer.from(String(a), "utf8");
  const bufB = Buffer.from(String(b), "utf8");
  if (bufA.length !== bufB.length) return false;
  return timingSafeEqual(bufA, bufB);
}

/**
 * Pure decision function. No network calls.
 * hook: { headers?: Record<string,string> } or {} if none configured
 * incoming: { headers: Record<string,string>, secretKeyName: string, mutationRanBeforeCheck: boolean }
 *
 * Returns one of:
 *   "UNVERIFIABLE_NO_SECRET"    - nothing was ever provisioned to check against
 *   "REJECT_MISMATCH"           - the secret header did not match
 *   "REJECT_USED_BEFORE_CHECK"  - matched, but a mutation ran before the check
 *   "TRUSTED"                  - matched, and checked before any mutation
 */
export function classifyWebhookRequest(hook, incoming) {
  const hookHeaders = hook.headers || {};
  if (Object.keys(hookHeaders).length === 0) return "UNVERIFIABLE_NO_SECRET";

  const key = incoming.secretKeyName;
  const expected = hookHeaders[key];
  const actual = (incoming.headers || {})[key];
  if (expected == null || actual == null || !constantTimeEqual(expected, actual)) {
    return "REJECT_MISMATCH";
  }

  if (incoming.mutationRanBeforeCheck) return "REJECT_USED_BEFORE_CHECK";

  return "TRUSTED";
}

async function bc(method, path, body) {
  const res = await fetch(BASE + path.replace(/^\//, ""), {
    method,
    headers: { "X-Auth-Token": TOKEN, "Content-Type": "application/json", "Accept": "application/json" },
    body: body ? JSON.stringify(body) : undefined,
  });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  const text = await res.text();
  return text ? JSON.parse(text) : null;
}

async function allHooks() {
  const resp = await bc("GET", "/v3/hooks?limit=250");
  return resp?.data || [];
}

async function hooksMissingSecret() {
  return (await allHooks()).filter((h) => !h.headers || Object.keys(h.headers).length === 0);
}

async function provisionSecret(hookId, headerName) {
  const value = randomBytes(32).toString("hex");
  return bc("PUT", `/v3/hooks/${hookId}`, {
    headers: { [headerName]: value },
    is_active: true,
  });
}

export async function run() {
  let fixed = 0;
  for (const hook of await hooksMissingSecret()) {
    console.warn(
      `Hook ${hook.id} scope=${hook.scope} destination=${hook.destination} has no secret header. ${DRY_RUN ? "would provision" : "provisioning"}`
    );
    if (!DRY_RUN) await provisionSecret(hook.id, HEADER_NAME);
    fixed++;
  }
  console.log(`Done. ${fixed} hook(s) ${DRY_RUN ? "to provision" : "provisioned"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: webhook-payload-not-verified/python/verify_webhook_secret.py
"""Classify BigCommerce webhook requests so a handler never trusts an unverified payload.

BigCommerce webhook callbacks carry a hash field, but BigCommerce has never published
a supported formula for validating it, so its presence is not real verification. The
documented safeguard is the optional headers object you set when creating a hook with
POST /v3/hooks, which BigCommerce echoes back on every callback as a shared secret.
This scans hooks for a missing secret, provisions one when confirmed missing, and
exposes a pure classification function a receiver can use before ever acting on the
payload. Run the scan on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/bigcommerce/webhook-payload-not-verified/
"""
import os
import hmac
import secrets
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("verify_webhook_secret")

STORE_HASH = os.environ["BIGCOMMERCE_STORE_HASH"]
TOKEN = os.environ["BIGCOMMERCE_ACCESS_TOKEN"]
BASE = f"https://api.bigcommerce.com/stores/{STORE_HASH}/"
HEADER_NAME = os.environ.get("WEBHOOK_SECRET_HEADER_NAME", "X-Webhook-Secret")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def bc(method, path, **kwargs):
    r = requests.request(
        method, BASE + path.lstrip("/"),
        headers={"X-Auth-Token": TOKEN, "Content-Type": "application/json", "Accept": "application/json"},
        timeout=30, **kwargs,
    )
    r.raise_for_status()
    return r.json() if r.content else None


def classify_webhook_request(hook, incoming):
    """Pure decision function. No network calls.

    hook: {"headers": {name: value, ...}} or {} / missing if none configured
    incoming: {"headers": {...}, "secretKeyName": str, "mutationRanBeforeCheck": bool}

    Returns one of:
      "UNVERIFIABLE_NO_SECRET"    - nothing was ever provisioned to check against
      "REJECT_MISMATCH"           - the secret header did not match
      "REJECT_USED_BEFORE_CHECK"  - matched, but a mutation ran before the check
      "TRUSTED"                  - matched, and checked before any mutation
    """
    hook_headers = hook.get("headers") or {}
    if not hook_headers:
        return "UNVERIFIABLE_NO_SECRET"

    key = incoming["secretKeyName"]
    expected = hook_headers.get(key)
    actual = (incoming.get("headers") or {}).get(key)
    if expected is None or actual is None or not hmac.compare_digest(expected, actual):
        return "REJECT_MISMATCH"

    if incoming.get("mutationRanBeforeCheck"):
        return "REJECT_USED_BEFORE_CHECK"

    return "TRUSTED"


def all_hooks():
    resp = bc("GET", "/v3/hooks?limit=250")
    return (resp or {}).get("data", [])


def hooks_missing_secret():
    """Hooks with no headers object at all, meaning nothing was ever provisioned to check."""
    return [h for h in all_hooks() if not h.get("headers")]


def provision_secret(hook_id, header_name):
    value = secrets.token_hex(32)
    return bc("PUT", f"/v3/hooks/{hook_id}", json={
        "headers": {header_name: value},
        "is_active": True,
    })


def run():
    fixed = 0
    for hook in hooks_missing_secret():
        log.warning(
            "Hook %s scope=%s destination=%s has no secret header. %s",
            hook.get("id"), hook.get("scope"), hook.get("destination"),
            "would provision" if DRY_RUN else "provisioning",
        )
        if not DRY_RUN:
            provision_secret(hook["id"], HEADER_NAME)
        fixed += 1
    log.info("Done. %d hook(s) %s.", fixed, "to provision" if DRY_RUN else "provisioned")


if __name__ == "__main__":
    run()
